proxy/context.go

```diff
@@ -15,48 +15,111 @@
 package proxy
 
 import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
 	"time"
 
+	"github.com/golang/glog"
 	"github.com/golang/protobuf/ptypes"
 	proxyconfig "istio.io/api/proxy/v1/config"
 	"istio.io/pilot/model"
 )
 
-// Context defines local proxy context information about the global service mesh
-type Context struct {
+// Environment provides an aggregate environmental API for Pilot
+type Environment struct {
 	// Discovery interface for listing services and instances
-	Discovery model.ServiceDiscovery
+	model.ServiceDiscovery
 
 	// Accounts interface for listing service accounts
-	Accounts model.ServiceAccounts
+	model.ServiceAccounts
 
 	// Config interface for listing routing rules
-	Config model.IstioConfigStore
+	model.IstioConfigStore
 
-	// MeshConfig defines global configuration settings
-	MeshConfig *proxyconfig.ProxyMeshConfig
+	// Mesh is the mesh config (to be merged into the config store)
+	Mesh *proxyconfig.ProxyMeshConfig
+}
+
+// Role declares the proxy node role in the mesh
+type Role interface {
+	// nolint: megacheck
+	isProxyRole()
+
```

Comment:
that's how subtyping works in golang, using a dummy method in a marker interface.